pkg/impersonation/impersonation.go

```diff
@@ -361,24 +361,45 @@ func (i *Impersonator) getUser(userInfo user.Info) (user.Info, error) {
 	}
 
 	groups := []string{"system:authenticated", "system:cattle:authenticated"}
+	extras := make(map[string][]string)
 	attribs, err := i.userAttributeLister.Get("", userInfo.GetUID())
 	if err != nil && !apierrors.IsNotFound(err) {
 		return &user.DefaultInfo{}, err
 	}
-	if attribs != nil {
+	if attribs == nil { // system users do not have userattributes, but principalid and username are on the user
```

Comment:
My concern with this if statement is that it links system users to not having userattributes at all. I don't know if we can take it for granted that this will always be true. In addition, if there's a change made which gives system users a user attribute but doesn't give a user attribute for these values (username/principal id) this code will fail.

One other way that you could do this would be to check if we found any username/ by going through the attributes, and if we didn't, use the values from the user. Below is what I am thinking:
```
if attribs != nil{
  // get attributes using current code
}
if usernames, ok := extras["username"]; !ok || len(usernames) == 0 {
    if u.Username != ""{
        extras["principalid"] = []string{u.PrincipalIDs}
    }else if u.DisplayName != ""{
        extras["username"] = []string{u.Username}
    }
}

if principalids, ok := extras["principalid"]; !ok || len(principalids) == 0{
    extras["principalid"] = u.PrincipalIDs
}
```
There's also the case where the attributes did not include the username/principalIDs which could be found on the user - for that case you'd have to loop through what you did find in the attribs and ensure that the username of the user and principalIDs from the user are in the overall master list. But that's a good bit of extra code/processing - interesting in hearing if you think it is necessary.

Reply:
> My concern with this if statement is that it links system users to not having userattributes at all. I don't know if we can take it for granted that this will always be true.

In my testing I found this to be the case. UserAttributes are only created when a real life user logs in and gets a token, and this does not happen for system users. @prachidamle can you comment on whether this is reliable?

> In addition, if there's a change made which gives system users a user attribute but doesn't give a user attribute for these values (username/principal id) this code will fail.

The code will not fail, it checks whether there are non-empty values for the username and principal ID. If they are not included in the UserAttributes, then they are not needed.

> One other way that you could do this would be to check if we found any username/ by going through the attributes, and if we didn't, use the values from the user
> There's also the case where the attributes did not include the username/principalIDs which could be found on the user 

If UserAttributes exists, it will be more complete than what is on the User because it will include data from the auth provider. There is no need to do extra processing.

Reply:
> Let's say that there is one user attribute created for a system user 

These aren't created for system users. They are created when a real user logs in: https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/tokens/manager.go#L648-L659

> Suppose that this attribute doesn't have the principal Id or the username in it. Since we got a user attribute back, this code won't use the values that are present for the system user on the user object.

The UserAttributes object, if it exists, is the authoritative source for user attributes, since it includes data from the auth provider.

> Is there anything that makes that guarantee

Rancher always adds the principal ID and username for every auth provider:

https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/providers/github/github_provider.go#L412-L421
https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/providers/saml/saml_provider.go#L399-L408
https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/providers/ldap/ldap_provider.go#L373-L382

Reply:
Changed this to only look for DisplayName for the `username` extra field, resolving

Reply:
I think this concern is already resolved, but yes the UserAttributes are only created when a real life user logs in and a token is created for that user. For system users having local auth, the UserAttributes are not created.